smack-extensions/src/main/java/org/jivesoftware/smackx/softwareinfo/element/SoftwareInfoElement.java

```diff
@@ -0,0 +1,219 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.softwareinfo.element;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.util.EqualsUtil;
+import org.jivesoftware.smack.util.HashCode;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+import org.jivesoftware.smackx.mediaelement.element.MediaElement;
+import org.jivesoftware.smackx.xdata.FormField;
+import org.jivesoftware.smackx.xdata.packet.DataForm;
+import org.jivesoftware.smackx.xdata.packet.DataForm.Type;
+
+/**
+ * {@link ExtensionElement} that contains Software Information. <br>
+ * Instance of SoftwareInfoElement can be created using {@link Builder#build()} method.
+ * To obtain Software Information use {@link #getForm()}.getFields()
+ */
+public final class SoftwareInfoElement implements ExtensionElement {
+
+    public static final String FORMTYPEVALUE = "urn:xmpp:dataforms:softwareinfo";
+    public static final String NAMESPACE = "jabber:x:data";
+    public static final String ELEMENT = "x";
+    private final DataForm form;
+    private final Map<String, FormField> fields;
+
+    private SoftwareInfoElement(Map<String, FormField> formFields) {
+        form = new DataForm(Type.result);
+        form.addFields(formFields.values());
+        this.fields = formFields;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    /**
+     * The DataForm returned contains all extension elements as {@link FormField}s.
+     * <br>
+     * Use dataform.getFields() to obtain Software Information.
+     * @return {@link DataForm}
+     */
+    public DataForm getForm() {
+        return form;
+    }
+
+    @Override
+    public int hashCode() {
+        HashCode.Builder builder = HashCode.builder();
+        for (FormField field : fields.values()) {
+            builder.append(field.getValues());
+        }
+        return builder.build();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return EqualsUtil.equals(this, obj, (equalsBuilder, otherObj) ->
+                                                                  equalsBuilder.append(fields.get("HEADER"), otherObj.fields.get("HEADER"))
```

Comment:
The level of indentation is to high. It would be more readable if you moved it a bit to the left.